File: solutions/7kyu/unique_string_characters.rb
# Задача на проверку уникальных букв в двух строках

# Разбить обе строки на массив и удалить элементы совпадающие по регулярке с другим масивом (метод delete_if),
# объединить и обратно в строки

def solve(a, b)
  ((a.split('').delete_if { |v| v =~ /["#{b}"]/ }) + (b.split('').delete_if { |v| v =~ /["#{a}"]/ })).join('')
end


# BEST solution
# Метод delete https://ruby-doc.org/core-2.7.1/String.html#method-i-delete
# no comment

# def solve(a,b)
#   a.delete(b) + b.delete(a)
# end

p solve("xyab", "xzca") #ybzc"
p solve("xyabb", "xzca") #"ybbzc"

File: solutions/7kyu/unique_string_characters.js
//Задача на проверку уникальных букв в двух строках
//solve("xyab","xzca") = "ybzc"

// Для решения использовалась функция filter, которая принимает в качестве аргумента callback функцию
// https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/filter
//

function solve(a,b){
  let filteredA = a.split('').filter(
      function(e) {
        return this.indexOf(e) < 0;
      },
      b.split('')
  );

  let filteredB = b.split('').filter(
      function(e) {
        return this.indexOf(e) < 0;
      },
      a.split('')
  );
  return filteredA.join('') + filteredB.join('');
}

// BEST solution
// Так же через filter c callback и стрелочной функцией,
// где проверяется вхождение фильтруемого элемента в строку a или строку b,
// через функцию
// function solve(a,b){
//  return (a+b).split("").filter(c => !a.includes(c) || !b.includes(c)).join("");
// }

console.log(solve("xyab", "xzca")); //ybzc"
console.log(solve("xyabb", "xzca")); //"ybbzc"